src/blk/kernel/KernelDevice.cc

```diff
@@ -731,6 +738,26 @@ int KernelDevice::queue_discard(interval_set<uint64_t> &to_release)
   return 0;
 }
 
+// return true only if queue_discard succeeded, so caller won't have to do alloc->release
+// otherwise false
+bool KernelDevice::try_discard_async_sync(interval_set<uint64_t> &to_release)
+{
+  if (!support_discard || !cct->_conf->bdev_enable_discard)
+    return false;
+
+  int r = 0;
+  if (cct->_conf->bdev_async_discard) {
```

Comment:
IMO better replace with
if (!discard_thread.is_started())

to handle the case when bdev_enable_discard has been updated on the fly